tutorials/MLlib.md

```diff
@@ -0,0 +1,268 @@
+1	1	[5.699999809265137, 2.5, 5.0, 2.0]	                [0.0, 39.0, 0.0, 1.0]	[0.0, 0.975, 0.0, 0.025]	    1.0
+2	1	[5.800000190734863, 2.700000047683716, 5.09999...	[0.0, 39.0, 0.0, 1.0]	[0.0, 0.975, 0.0, 0.025]	    1.0
+3	1	[6.099999904632568, 2.5999999046325684, 5.5999...	[0.0, 3.0, 0.0, 2.0]	[0.0, 0.6, 0.0, 0.4]	            1.0
+4	1	[6.800000190734863, 3.200000047683716, 5.90000...	[0.0, 39.0, 0.0, 1.0]	[0.0, 0.975, 0.0, 0.025]	    1.0
+5	1	[7.199999809265137, 3.5999999046325684, 6.0999...	[0.0, 39.0, 0.0, 1.0]	[0.0, 0.975, 0.0, 0.025]	    1.0
+6	1	[7.599999904632568, 3.0, 6.599999904632568, 2....	[0.0, 39.0, 0.0, 1.0]	[0.0, 0.975, 0.0, 0.025]	    1.0
+7	2	[5.0, 3.5999999046325684, 1.399999976158142, 0...	[0.0, 0.0, 49.0, 0.0]	[0.0, 0.0, 1.0, 0.0]	            2.0
+8	3	[4.900000095367432, 2.4000000953674316, 3.2999...	[0.0, 1.0, 0.0, 41.0]	[0.0, 0.023809523809523808...	    3.0
+9	3	[5.5, 2.299999952316284, 4.0, 1.2999999523162842]	[0.0, 1.0, 0.0, 41.0]	[0.0, 0.023809523809523808...	    3.0
+```
+
+### Klastrowanie
+
+Do klastrowania danych możemy użyć algorytmu k-średnich i klasy `KMeans`. 
+W pakiecie MLlib znajdują się również metody ewaluacji modeli jak np `ClusteringEvaluator`.
+
+```python
+from pyspark.ml.evaluation import ClusteringEvaluator
+from pyspark.ml.clustering import KMeans
+
+kmeans = KMeans().setK(3)
+model = kmeans.fit(train)
+
+predictions = model.transform(train)
+
+evaluator = ClusteringEvaluator()
+silhouette = evaluator.evaluate(predictions)
+print("Silhouette with squared euclidean distance = " + str(silhouette))
+
+model.summary.predictions.show(5)
+```
+
+```commandline
+Silhouette with squared euclidean distance = 0.7522124869541736
++------+--------------------+----------+
+|labels|            features|prediction|
++------+--------------------+----------+
+|     1|[4.90000009536743...|         1|
+|     1|[5.80000019073486...|         1|
+|     1|[5.80000019073486...|         1|
+|     1|[5.90000009536743...|         1|
+|     1|[6.0,2.2000000476...|         1|
++------+--------------------+----------+
+only showing top 5 rows
+```
+
+### PCA
+Do zmniejszenia liczby wymiarów możemy użyć algorytmu PCA. Wywołanie 
+przebiega następująco:
+
+```python
+from pyspark.ml.feature import PCA
+
+pca = PCA(k=2, inputCol="features")
+pca.setOutputCol("pca_features")
+model = pca.fit(train)
+
+features = model.transform(train)
+features.show(5)
+```
+
+```commandline
++------+--------------------+--------------------+
+|labels|            features|        pca_features|
++------+--------------------+--------------------+
+|     1|[4.90000009536743...|[-6.0155091733474...|
+|     1|[5.80000019073486...|[-6.9085974889527...|
+|     1|[5.80000019073486...|[-7.0791582734907...|
+|     1|[5.90000009536743...|[-6.8828186812519...|
+|     1|[6.0,2.2000000476...|[-6.7950916056033...|
++------+--------------------+--------------------+
+only showing top 5 rows
+```
```

Comment:
Brakuje mi funkcji, które już użyliśmy w projekcie, tj.: StringIndexer, OneHotEncoder, SparseVector, Imputer. Chociaż nie wiem, czy są konieczne. @PiotrMaciejKowalski opinia?

Reply:
Tu nie jestem pewien czy dodawać coś co mamy zaimplementowane wewnątrz naszych funkcji, no bo raczej nie będziemy już ruszać np metody OneHotEncoder

Reply:
Nie to byłoby wsteczne. Tutorial powinien być niezależny od projektu.